### DISCUSSION TITLE
What is orphans and why number of orphans still growing

### DISCUSSION THREAD

**Discussion by mag101** [seaweedfs/seaweedfs#4636] [General]
Sponsors SeaweedFS via Patreon https://www.patreon.com/seaweedfs

Please describe what is orphans and from where  appear.
Number of orphans is growing discontinue. They occupy much more space then really data.
This problem reproducing on more the 150 different installations. 
seaweedfs version 3.40
we use next installations:
fuse+weed -> 3 masters -> two fillers -> two volumes servers.

**Comment by chrislusf** [comment]
please describe your use cases.

**Comment by mag101** [comment]
we use cluster solution 
![image](https://github.com/seaweedfs/seaweedfs/assets/40891107/3f09fc28-3364-4609-b5c3-3c042c5f6fb7)

**Reply by chrislusf** [reply]
do not put a proxy between mount and filer, or at least keep the connection between mount and filer sticky.
they need a streaming grpc connection between them.

**Comment by mag101** [comment]
We have half of installation where haproxy use and another half without proxy with direct mount. Result the same - orphans still growing.  We use multiple mount points from number VMs in a cluster  in additional to this diagram. This VMs creating/deleting files and directories. Is it possible in our version    3.40 not implemented  [mount] fix metacache update #4161 yet and files going to orphans?

**Reply by chrislusf** [reply]
could be related to this https://github.com/seaweedfs/seaweedfs-csi-driver/issues/127#issuecomment-1625671846

Need to create the directory on filer before using mounting to it.

**Comment by mag101** [comment]
Why I ask about orphans  - it impact service:
after some time for one installation days another weeks ( depends haw often files changing )  we have volume full of orphans and service fail. And only one possible to fix clear orphans with command echo -e "lock\nvolume.fsck -reallyDeleteFromVolume\nunlock | weed shell" and then curl "http://localhost:9333/vol/vacuum?garbageThreshold=0.005" .
Used space decrease in 4-5 times. According log vacuum is starting automatically with default Threshold.

**Comment by wusongANKANG** [comment]
Orphans are chunks that are in the volume but not in the metadata.

1: All deleted data but not garbage collected are orphans.
2: At the same time, when an io error occurs when uploading data, there will also be orphans.

So the orphans are always growing before the vacuum is executed

**Comment by mag101** [comment]
Thank a lot  for clarify. Is it safe to clear orphans by  echo -e "lock\nvolume.fsck -reallyDeleteFromVolume\nunlock | weed shell" ?
Without this command vacuum doesn't help  with garbageThreshold value  less then default 0.5 yet.

**Comment by mag101** [comment]
Another questions in this case : How to properly monitor orphans issue?
when orphans still growing we can check free space on seaweedfs  by command
curl -m -2 -s localhost:8080/status?pretty=y | jq .DiskStatuses[].percent_free
If free space percent less then 30% we have issues with mounts. The same time  no any readonly volumes found.
Problem is in this check: some time it's correct but another time  we have issues with free space between 30-35%.
After clean orphans and vacuum free space  ~ 90-95%
Is it additional or another parameters that we need to monitor and have exactly result in advance to prevent space issues?
Thanks for any answer

**Comment by kmlebedev** [comment]
If the orphans grow, then something goes wrong and this should be written in the logs, since it is expected that the data is deleted / overwritten sequentially in meta then the blob

**Comment by mag101** [comment]
No any issues  in logs found. Is it possible issues with filer sync, how filer sync sensitive to number of chunks?
In our deployment two filers in the same google network but in different google zones for HA.

**Comment by elesueur** [comment]
@mag101 we too are finding that orphans are accumulating. Our architecture for SeaweedFS is fairly simple, single master, single filer, 3 volume servers. These run in kubernetes and are spread across 3 nodes (VMs) that operate on hosts (ESXi) in the same datacenter rack.

I ran `volume.fsck -v` and interestingly, when requesting for those chunks to the corresponding volume server, it returns a 404. I would have expected the chunk to still be readable if it is reported as an orphan. Even with `?readDeleted=true` on the request, I still get a 404 error.

What did you end up doing to manage the cluster and its orphan accumulation?

**Reply by chrislusf** [reply]
> @mag101 we too are finding that orphans are accumulating. Our architecture for SeaweedFS is fairly simple, single master, single filer, 3 volume servers. These run in kubernetes and are spread across 3 nodes (VMs) that operate on hosts (ESXi) in the same datacenter rack.
> 
> I ran `volume.fsck -v` and interestingly, when requesting for those chunks to the corresponding volume server, it returns a 404. I would have expected the chunk to still be readable if it is reported as an orphan. Even with `?readDeleted=true` on the request, I still get a 404 error.
> 
> What did you end up doing to manage the cluster and its orphan accumulation?

* Your case seems losing the chunk data instead of orphans. 
* Try to make a reproducible case.

**Comment by ginkel** [comment]
Hi there,

my SeaweedFS (currently v3.85) cluster seems to have accumulated a substantial mount of orphans - I hope it is ok if I append my case to this thread:

```
$ mcli du seaweedfs/   
115GiB  2335892 objects

[...]

> volume.fsck
[...]
Total           entries:7999340 orphan:3277330  40.97%  319035846484B
This could be normal if multiple filers or no filers are used.
```

I use a (Docker-based) three-node architecture with three masters, three filers and three volume servers w/ the default filer configuration, i.e. level2db.

Are these orphans emitted by `volume.fsck` false positives due to the fact that I run three filers?

I don't think so as `du` on the volume servers show way more disk space being used for SeaweedFS volumes on a single node than the cluster hosts data:

```
root@pegasus - /srv/seaweedfs/volume # du -skh  
234G    .

root@deneb - /srv/seaweedfs/volume # du -skh
307G    .

root@antares - /srv/seaweedfs/volume # du -skh
358G    .
```

`volume.vacuum` did not reclaim a substantial amount of disk space.

Any ideas how to further troubleshoot this?

Thanks,
Thilo

**Comment by chrislusf** [comment]
`volume.fsck -reallyDeleteFromVolume` can delete orphan data.

**Reply by ginkel** [reply]
Thanks! I tried this a couple of times. The result does not seem to be exhaustive, though, i.e., multiple runs of `volume.fsck -reallyDeleteFromVolume` delete more orphans on each run. The overall space usage has improved a bit, though, but is still more inefficient than expected:

```
root@pegasus - /srv/seaweedfs/volume # du -skh
185G    .

# mcli du seaweedfs/ 
120GiB  2777973 objects
 ```

Any further ideas?

Thanks,
Thilo

**Reply by chrislusf** [reply]
after deleting from volume servers `volume.fsck -reallyDeleteFromVolume`, to reduce volume size, need to vacuum the volumes.

**Reply by ginkel** [reply]
Done. I also made sure all empty buckets are deleted, which freed another good amount of storage. Thanks!
